### ISSUE TITLE
Slow eslint rules

### ISSUE THREAD

**Issue by nishantwrp** [typescript-eslint/typescript-eslint#2450]
This is a problem with rules like `@typescript-eslint/no-floating-promises` that require `parserServices` to be generated (and require "project" property to be set). Without this rule, the eslint checks take around `40min` to run completely. But when I add a rule that requires `parserServices` and set the "project" property in my eslint config the time eslint takes jumps to around `4hrs`. Is this expected?

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#my-linting-feels-really-slow

Also some discussion on this issue: https://github.com/typescript-eslint/typescript-eslint/issues/1192

TL;DR:
- type information takes time to generate. That time is directly proportional to the size of your project.
- type information is mostly computed lazily, the more types that are accessed, the more work that needs to be done.

The rules aren't perfect either. Sometimes they do work they don't need to and ask for types they don't need. They can always be optimised - happy to accept a PR!

**Comment by nishantwrp** [comment]
Got it @bradzacher.
